### PR TITLE
dnsdist: allow setting keyLogFile to DoT/DoH backends

### PR DESCRIPTION
**Pull request by karelbilek**
### Short description

This adds keyLogFile setting to DoT and DoH backends.

### Checklist

I have:
- [x] read the [CONTRIBUTING.md](https://github.com/PowerDNS/pdns/blob/master/CONTRIBUTING.md) document
- [x] compiled this code
- [x] tested this code - tested with wireshark + google's DoH backend
- [x] included documentation (including possible behaviour changes)
- [ ] documented the code
- [x] added or modified regression test(s)
- [ ] added or modified unit test(s)

The regression tests just try if the logfiles exist and are non-empty, they don't really check what is in them.

### CODE REVIEW

File: pdns/tcpiohandler.cc [hunk 1]
```diff
@@ -951,6 +955,7 @@ class OpenSSLTLSIOCtx: public TLSCtx, public std::enable_shared_from_this<OpenSS
   std::shared_ptr<SSL_CTX> d_tlsCtx{nullptr}; // client context, on a server-side the context is stored in d_feContext->d_tlsCtx
   std::unique_ptr<OpenSSLFrontendContext> d_feContext{nullptr};
   bool d_ktls{false};
+  pdns::UniqueFilePtr d_keyLogFile{nullptr};
```
Comment (rgacogne) [inline pdns/tcpiohandler.cc:958]: Just a nit, please move the pointer above the boolean so we don't waste space due to padding:
```suggestion
  pdns::UniqueFilePtr d_keyLogFile{nullptr};
  bool d_ktls{false};
```

File: pdns/dnsdistdist/dnsdist-lua.cc [hunk 2]
```diff
@@ -559,6 +559,15 @@ static void setupLuaConfig(LuaContext& luaCtx, bool client, bool configCheck)
                          getOptionalValue<bool>(vars, "ktls", config.d_tlsParams.d_ktls);
                          getOptionalValue<std::string>(vars, "subjectName", config.d_tlsSubjectName);
 
+                         if (vars->count("keyLogFile") > 0) {
+#ifdef HAVE_SSL_CTX_SET_KEYLOG_CALLBACK
+                           getOptionalValue<std::string>(vars, "keyLogFile", config.d_tlsParams.d_keyLogFile);
+#else
+                           errlog("TLS Key logging has been enabled using the 'keyLogFile' parameter to newServer(), but this version of OpenSSL does not support it");
```
Comment (karelbilek) [inline pdns/dnsdistdist/dnsdist-lua.cc:566]: I don't understand why code formatter doesn't like this line

File: pdns/tcpiohandler.hh [hunk 3]
```diff
@@ -587,6 +587,7 @@ struct TLSContextParameters
   bool d_releaseBuffers{true};
   bool d_enableRenegotiation{false};
   bool d_ktls{false};
+  std::string d_keyLogFile;
```
Comment (rgacogne) [inline pdns/tcpiohandler.hh:590]: Same here, please move it just below `d_caStore`.

**Review by rgacogne — CHANGES_REQUESTED** [review]
The PR looks good to me, I requested two very small changes for consistency's sake. Thanks a lot!

File: pdns/dnsdistdist/dnsdist-lua.cc (hunk 2 above)
Comment (rgacogne) [reply pdns/dnsdistdist/dnsdist-lua.cc]: It's clearly a bug in `clang-format`, I suggest you ignore it.

Comment (karelbilek) [reply pdns/dnsdistdist/dnsdist-lua.cc]: Ah, I reformatted the file to make clang-format happy.

Comment (rgacogne) [reply pdns/dnsdistdist/dnsdist-lua.cc]: That's fine, I can reformat later if it annoys me too much :)

**Comment by karelbilek** [comment]
I haven't been able to run the regression tests locally so far, and they don't seem to be run in CIs on PR... should I try more? :)

**Comment by rgacogne** [comment]
> I haven't been able to run the regression tests locally so far, and they don't seem to be run in CIs on PR... should I try more? :)

They are, what makes you think otherwise?

**Comment by rgacogne** [comment]
(look for `Build and test everything / test-dnsdist-regression`)

**Review by rgacogne — APPROVED** [review]
Looks great, thanks!

**Comment by karelbilek** [comment]
@rgacogne just a thought, should I backport it to 1.9? (locally I actually run it with 1.9, it works fine)

**Comment by rgacogne** [comment]
At this point I would prefer not backporting it: introducing new features in stable branchs is always risky, this one touches code that is quite critical (even though the change is very simple and clean), I'm not sure adding it's worth the risk. If someone really needs it to investigate an issue it's not hard for us to build custom packages backporting the feature.

**Comment by karelbilek** [comment]
Okay!

**Pull request merged**